articles/ros2_security_contexts.md

```diff
@@ -0,0 +1,307 @@
+
+
+### Modeling permissions of nodes in a process v.s. permission of the middleware ``Participant``
+
+Before the use of contexts, multiple nodes composed into a single process where each mapped to a separate ``Participant``.
+Each ``Participant`` subsequently load a security identity and access control credential prevalent to its' respective node.
+However, all nodes in that process share the same memory space and can thus access data from other nodes.
+There is a mismatch between the middleware credentials/permissions loaded and the resources accessible within the process.
+
+By using contexts, all nodes in a context share the same security identity and access control credentials.
+This inevitably means that code compiled to node ``foo`` can access credentials/permissions only trusted to node ``bar``.
+This consequence of composition could unintendedly subvert the minimal spanning policy as architected by the policy designer or measured/generated via ROS 2 tooling/IDL.
+
+With the introduction of contexts, it becomes possible to describe the union of access control permission by defining a collection of SROS 2 policy profiles as element within a specific context.
+This would allow for formal analysis tooling [2] to check for potential violations in information flow control given the composing of nodes at runtime.
+If a process contains a single context, this reconciles the permissions of a ``Participant`` and the ones of the process.
+
+However, should multiple contexts be used per process, then such security guaranties are again lost because both contexts will share the same memory space.
+Thus it should be asked whether if multiple contexts per process should even be supported.
+
+
+In summary, the distinction here is that before, the composition of multiple permissions could not be conveyed to the tooling.
+Whether nodes could gain the permission of others in the same process space is not the hinge point of note; it's the fact that such side effects could not be formally modeled or accounted for by the designer.
+It will now be possible with contexts, however allowing for multiple contexts per process will reintroduce and exacerbates the same modeling inaccuracies.
+
+### Composable launchfile includes
+
+A particular challenge in using launchfiles with security contexts is that of keeping the include hierarchy composable.
+An inherit tradeoff between simplicity and configurability can arise when writing launchfiles for downstream use.
+Authors can selectively choose what attributes to expose as input arguments, while users may implicitly override provided defaults.
+
+In case of contexts, it is not inherently clear what best practices either package authors or users should employ to retain a composable and intuitive launchfile structure. E.g:
+Should authors parametrize context paths for each node as input arguments?
```

Comment:
IMO, yes

Reply:
@mikaelarguedas and I where wondering if that would make things too noisy or add to the boilerplate in launchfiles. Given node name uniqueness, we could encourage users to just name there relative context the same as the node name by convention, and rely on pushing namespaces. (I guess that gets us back to why not just default the relative context name being the node name for single process nodes)

The only deviation would be when composing nodes into containers. Is there an example we can look at on how one may create a container in a parent launchfile, then include a child launchfile by passing the container (reference/name/or-what?) its' nodes are expected to compose into.

Reply:
> I guess that gets us back to why not just default the relative context name being the node name for single process nodes.

I think that that is a good idea. See my other comment: https://github.com/ros2/design/pull/274#issuecomment-592071949.

Reply:
> The only deviation would be when composing nodes into containers. Is there an example we can look at on how one may create a container in a parent launchfile, then include a child launchfile by passing the container (reference/name/or-what?) its' nodes are expected to compose into.

I can think there are a few ways of achieving that, but I don't think we have a recommended way of doing it.

Reply:
Statement formalized here https://github.com/ros2/design/pull/274/files#diff-0c022ad3d42831e4e8cf2336da70d354R240